File: SecondYear/sauban123_MohdSauban_2226csit1130_2/Week_4/MinimumParenthesis.cpp
#include <bits/stdc++.h> 
int minimumParentheses(string pattern) {
    int count = 0;
    int open = 0;
    for(int i=0;i<pattern.length();i++)
    {
        if(pattern[i]==')' && open<=0)
        {
            count++;
        }
        else if(pattern[i]==')' && open>0)
        {
            open--;
        }
        else if(pattern[i]=='(')
        {
            open++;
        }
    }
    return count+open;
}

File: SecondYear/aayushraghav93_AayushRaghav_2226CSE(AI)1108_2/Week_4/minimumparenthesis.java
import java.util.* ;
import java.io.*; 
public class Solution {
	public static int minimumParentheses(String pattern) {
		if (pattern == null || pattern.length() == 0) {
            return 0;
        }
        Stack<Character> stack = new Stack<>();
        int count = 0;
        for (char c : pattern.toCharArray()) {
            if (c == '(') stack.push(c);
            else if (stack.isEmpty()) count++;
            else stack.pop();
        }
        return count + stack.size();
	}
}